### PR TITLE
Added a blas package.

### PR DESCRIPTION
**Pull request by pelson**
Ping @jakirkham. I will build out the matrix once we have a feedstock (`openblas` & `noblas`).

### CODE REVIEW

**Comment by pelson** [comment]
Seems the linter can't yet deal with `os.environ` - I thought I'd dealt with that, but apparently not. I wont try and fix it now though. One for the TODO list.

File: recipes/blas/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Some useful notes can be found at https://conda-forge.hackpad.com/BLAS-Numpy-Friends-86De62hoNdT
+
+{% set variant = os.environ.get('BLAS_VARIANT', 'openblas') %}
```
Comment (jakirkham) [inline recipes/blas/meta.yaml]: 👍

File: recipes/blas/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+build:
+  # We mustn't use the build number in the blas package - because we are defining the string,
+  # manually we would not be producing unique filenames.
+  number: 0
+  string: {{ variant }}
+  track_features:
+    # The "blas" package must always track a feature, else it will be
+    # prioritised over all others.
+    - blas_{{ variant }}
```
Comment (jakirkham) [inline recipes/blas/meta.yaml]: I think we need to add this to the `build` section for the environment variable to work.

```
script_env:
  - BLAS_VARIANT
```

**Comment by jakirkham** [comment]
Just a couple thoughts.

* Probably need to specify `BLAS_VARIANT` for pass through.
* Would it be worth including our notes on BLAS as a Readme.md?

File: recipes/blas/meta.yaml (hunk 2 above)
Comment (pelson) [inline recipes/blas/meta.yaml]: No, I haven't found that to be the case. `BLAS_VARIANT=foobar conda build recipes/blas` produces the expected results. The variable *would* be needed in `script_env` if we wanted access to it in `build.sh`.

**Comment by pelson** [comment]
> Would it be worth including our notes on BLAS as a Readme.md in the recipe directory?

I'm happy to do so, but honestly I think we are just as well keeping them in one place on the hackpad. Your call

File: recipes/blas/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/blas/meta.yaml]: Yep, you are right this works with `jinja`. I can't really think of a reason to have a `build.sh` so let's leave it.

**Comment by jakirkham** [comment]
> I'm happy to do so, but honestly I think we are just as well keeping them in one place on the hackpad. Your call

Yeah, you're right. We may need to tweak things quickly. PRs for this would be mildly annoying.

How about this, we add a Readme.md that just has a link to the hackpad. In the future, we can move things in when we are done.

<br>
In any event, I'm quite happy to merge before doing any of this. It would better if we can work on the next stage. Sound good?

**Comment by pelson** [comment]
> In any event, I'm quite happy to merge before doing any of this. It would better if we can work on the next stage. Sound good?

Go, go, go! 👍

**Comment by jakirkham** [comment]
Gone. 🐦

**Comment by jakirkham** [comment]
Please let me know if there is anything I can do to help with getting the rest of this done.

**Comment by jakirkham** [comment]
FYI this is now released on all platforms if you are wanting to experiment more.

**Comment by scopatz** [comment]
Hi @pelson @jakirkham! Thanks for putting this in!  One thing that has been confusing me, probably because I haven't followed it closely enough, is how another package is supposed to select a variant.  For instance, this is an issue with the CoinCbc package, that currently depends on openblas.

**Comment by jakirkham** [comment]
The idea that was discussed (though I don't believe this exists *yet*) is that we would have matrix builds for variants. Right now the only variants considered are not having a BLAS at all and using OpenBLAS. The former is important for memory constrained scenarios (e.g. Heroku).

Other things we might consider in the future include ATLAS and Blis. We are not sure if ATLAS can build on the CIs so it might be out. Also, it's lack of runtime selection is a huge downside. I know the NumPy folks have a way of working around this, but it sounds challenging to implement.

Important side note: We haven't figured out how to build OpenBLAS on CI within the time constraints in an acceptable manner yet. Just something to keep in mind.

In short, stay tuned for more updates.

**Comment by scopatz** [comment]
Ok thanks @jakirkham!

**Comment by pelson** [comment]
To give a specific answer, you make `blas 1.0 openblas` a runtime dependency of the package, and list the `blas_openblas` feature. It is the best we could come up with which behaves reasonably well with the existing nomkl / mkl stuff in defaults and which gives us proper blas variant control.

**Comment by pelson** [comment]
> We haven't figured out how to build OpenBLAS on CI within the time constraints in an acceptable manner yet

@jakirkham: https://github.com/conda-forge/openblas-feedstock and https://anaconda.org/conda-forge/openblas? I'm clearly missing some detail... 😕

**Comment by jakirkham** [comment]
> I'm clearly missing some detail... 😕

Yes, the detail is on AppVeyor. We were hitting the 1hr mark before. Though we were also installing MSYS2 by hand. I'm hoping if we use @mingwandroid's packages that should help some. Maybe not enough though.

**Comment by pelson** [comment]
Maybe we can dodge that bullet by copying something like the openblases here https://anaconda.org/mutirri/openblas temporarily.

**Comment by jakirkham** [comment]
I'm not loving the idea of pulling this from someone else's channel. Would much rather skip building NumPy on Windows with OpenBLAS at present.

**Comment by jakirkham** [comment]
Though I would be tempted to get it via a conda MSYS2 package for the near term if that is agreeable.
